pkg/util/apply/apply.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2025 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	corev1ac "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/utils/ptr"
+)
+
+func UpsertEnvVar(envVarList *[]corev1ac.EnvVarApplyConfiguration, envVar ...*corev1ac.EnvVarApplyConfiguration) {
+	for _, e := range envVar {
+		upsert(envVarList, *e, byEnvVarName)
+	}
+}
+
+func UpsertEnvVars(envVarList *[]corev1ac.EnvVarApplyConfiguration, envVars []corev1ac.EnvVarApplyConfiguration) {
+	for _, e := range envVars {
+		upsert(envVarList, e, byEnvVarName)
+	}
+}
+
+func UpsertPort(portList *[]corev1ac.ContainerPortApplyConfiguration, port ...*corev1ac.ContainerPortApplyConfiguration) {
+	for _, p := range port {
+		upsert(portList, *p, byContainerPortOrName)
+	}
+}
+
+func UpsertVolumes(volumeList *[]corev1ac.VolumeApplyConfiguration, volumes []corev1ac.VolumeApplyConfiguration) {
+	for _, v := range volumes {
+		upsert(volumeList, v, byVolumeName)
+	}
+}
+
+func UpsertVolumeMounts(mountList *[]corev1ac.VolumeMountApplyConfiguration, mounts []corev1ac.VolumeMountApplyConfiguration) {
+	for _, m := range mounts {
+		upsert(mountList, m, byVolumeMountName)
+	}
+}
+
+func byEnvVarName(a, b corev1ac.EnvVarApplyConfiguration) bool {
+	return ptr.Equal(a.Name, b.Name)
+}
+
+func byContainerPortOrName(a, b corev1ac.ContainerPortApplyConfiguration) bool {
+	return ptr.Equal(a.ContainerPort, b.ContainerPort) || ptr.Equal(a.Name, b.Name)
+}
+
+func byVolumeName(a, b corev1ac.VolumeApplyConfiguration) bool {
+	return ptr.Equal(a.Name, b.Name)
+}
+
+func byVolumeMountName(a, b corev1ac.VolumeMountApplyConfiguration) bool {
+	return ptr.Equal(a.Name, b.Name)
+}
+
+type compare[T any] func(T, T) bool
```

Comment:
```suggestion
type compare func(any, any) bool
```

Is this type parameter needed?

Reply:
I think that's needed to constrain both variables have the same type.

Reply:
Interesting, thank you for checking!
Once k/k provides interface for ApplyConfiguration, we can just use that, right?

Reply:
I'm not sure, generics are not covariant, so even if you have an common interface, that doesn't mean it's possible to compare different types implementing that interface. That's similar to array that are covariants while generics are invariants.

Reply:
Yeah, that makes sense. Generics is not typed, just reference to multiple typed.